internal/vmio/vsock.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package vmio
+
+import (
+	"context"
+	"io"
+	"net"
+	"time"
+
+	"github.com/mdlayher/vsock"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// VSockDial attempts to connect to a vsock listener at the provided cid and port with a hardcoded number
+// of retries.
+func VSockDial(reqCtx context.Context, logger *logrus.Entry, contextID, port uint32) (net.Conn, error) {
+	// VM should start within 200ms, vsock dial will make retries at 100ms, 200ms, 400ms, 800ms, 1.6s, 3.2s, 6.4s
+	const (
+		retryCount      = 7
+		initialDelay    = 100 * time.Millisecond
+		delayMultiplier = 2
+	)
+
+	var lastErr error
+	var currentDelay = initialDelay
+
+	for i := 1; i <= retryCount; i++ {
+		select {
+		case <-reqCtx.Done():
+			return nil, reqCtx.Err()
+		default:
+			conn, err := vsock.Dial(contextID, port)
+			if err == nil {
+				logger.WithField("connection", conn).Debug("Dial succeeded")
+				return conn, nil
+			}
+
+			logger.WithError(err).Warnf("vsock dial failed (attempt %d of %d), will retry in %s", i, retryCount, currentDelay)
+			time.Sleep(currentDelay)
+
+			lastErr = err
+			currentDelay *= delayMultiplier
+		}
+	}
+
+	logger.WithError(lastErr).WithFields(logrus.Fields{"context_id": contextID, "port": port}).Error("vsock dial failed")
+	return nil, lastErr
+}
+
+// VSockDialConnector provides an IOConnector interface to the VSockDial function.
+func VSockDialConnector(contextID, port uint32) IOConnector {
+	return func(procCtx context.Context, logger *logrus.Entry, errG *errgroup.Group) <-chan io.ReadWriteCloser {
```

Comment:
I would use a struct if the parameters (`contextID` and `port`) needed to be accessed by different functions or multiple times, but they are only used by a single function for a single invocation.

If we made this a struct, AFAIK it would still be allocated on the heap because it is passed far out from the original function call (to other goroutines).

Reply:
Okay, that makes sense. And yes, the struct would also be allocated on the heap, but just once. This has two allocations, although minor it is good to know!